client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -48,7 +52,23 @@
 @ToString(of = { "stream", "closed" })
 public class EventStreamWriterImpl<Type> implements EventStreamWriter<Type> {
 
-    private final Object lock = new Object();
+    /**
+     * The lock here is used to enforce the following behaviour
+     *
+     * a. When a Write is happening, segmentSealedCallback cannot be executed concurrently, this is used to handle
+     * missing event.
+     * b. When a Write is happening, a newer write cannot be execute concurrently.
+     * c. When a Write is happening, flush cannot be executed concurrently.
+     * d. When a Flush is being invoked, segmentSealedCallback can be executed concurrently.
+     * e. When a Flush is being invoked, write cannot be executed concurrently.
+     * f. When a Close is being invoked, write cannot be executed concurrently.
+     * g. When a Close is being invoked, Flush and segmentSealedCallback can be executed concurrently.
+     */
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
```

Comment:
given that we are using a qualifier in `segmentSealedLock` below, should we also qualify this lock, something like `writeFlushLock`?